### ISSUE TITLE
Tab-complete

### ISSUE THREAD

**Issue by heyitsmeuralex** [towerofnix/bantisocial#56]
In message input, for:
* **Channel/server references**
  - `#channel` by channels on current server
  - `+server.com#channel` by fetching channels on remote server
  - `+server.com` by looking at known servers (i.e. present in server list)
* **Mentions**
  - `@username` by users on server (blocked by #55)

...anything else?

**Comment by heyitsmeuralex** [comment]
> what do we do with ambiguity?

I say we just abort - on Discord the tab-complete dropdown kinda lags the entire UI out, and I'd like to avoid that.

**Comment by towerofnix** [comment]
>I say we just abort - on Discord the tab-complete dropdown kinda lags the entire UI out, and I'd like to avoid that.

@heyitsmeuralex Sounds good to me. Maybe show the first three (or so) matches before aborting.

**Comment by heyitsmeuralex** [comment]
>Maybe show the first three (or so) matches

How? Design-wise, I mean.

**Comment by towerofnix** [comment]
If the actual processing of tab-completion doesn't take long (i.e. if the rendering of the DOM elements is what causes lag), I'd suggest just finding all completions, sorting alphabetically, and showing the first three.

This seems at least a little similar to how Discord does it actually? On a large (many members) server:

![dropdown](https://user-images.githubusercontent.com/9948030/33807978-b1e6ec82-ddd6-11e7-9716-0e731939f109.png)

Discord limits the number of users rendered. I have no idea *how* it's sorted (e.g. "Axi" appears after "Taylor", so it's not (just?) alphabetical). Point is - it's not that many elements being shown, and, frankly, Discord's tab-complete doesn't lag me at all.

Prioritizing online users may be a good idea.

**Comment by heyitsmeuralex** [comment]
That tab complete on Discord looks noticeably different to how I remember it - perhaps they updated/optimized it :tada:

Hitting tab on just `@` is a good question though. Should we even complete it?

**Comment by towerofnix** [comment]
I'm leaning towards "yes" :)

Prioritize users who are online, then sort alphabetically. I think that's enough?

**Comment by PullJosh** [comment]
No! Sort by who posted most recently in the current channel. ;)

**Comment by towerofnix** [comment]
@PullJosh Of course! That makes a lot more sense. In fact, oh, duh. :+1:

**Comment by TheInitializer** [comment]
Discord, and most other chat things, work by showing the people who sent the last chat messages first.

e.g.

**alex** we should add tab completion
**florrie** oh yeah, duh
**xn--cr8h** how should we do it?
**florrie** idk
**pulljosh** i can do frontend!

@ would show pulljosh, then florrie, then xn--cr8h.

**Comment by TheInitializer** [comment]
oops, ninja'd by @PullJosh 📦
